src/dmtcpworker.cpp

```diff
@@ -390,7 +390,7 @@ DmtcpWorker::waitForCheckpointRequest()
 
 // now user threads are stopped
```

Comment:
Most of my comments below can be ignored, if you agree to do the following refactoring.  I think it's simpler.

You used to have::
```
threadlist.cpp:checkpointhread():
  DmtcpWorker::preCheckpoint(); 
```

Right now, you have split `preCheckpoint` into `releaseLocks` and `preCheckpoint`:
```
threadlist.cpp:checkpointhread():
  DmtcpWorker::releaseLocks();
  ThreadList::waitForExitingThreads();
  DmtcpWorker::preCheckpoint(); 
```

But in this case, why not keep the old `preCheckpoint`, and simply insert waitForExitingThreads()` into the right location inside `preCheckpoint`:

```
  JTRACE("Threads suspended");
  WorkerState::setCurrentState(WorkerState::SUSPENDED);

  ThreadSync::releaseLocks();

  if (exitInProgress) {
    // There is no reason to continue checkpointing this process as it would
    // simply die right after resume/restore.
    // Release user threads from ckpt signal handler.
    ThreadList::resumeThreads();
    ckptThreadPerformExit();
  }

  ThreadList::waitForExitingThreads();

  ...
```

Reply:
Sorry for my delay.  I'll look at this carefully on Wednesday.
Probably the right thing, then, is to add a comment saying that long-term, we want ThreadList to call DmtcpWorker, and never vice versa.  I can look at this closely on Wednesday, and suggest a comment to document this (and maybe a strategy to prevent future entanglement in case someone again tries to call ThreadList from DmtcpWorker).

If you have a way of making future PRs to have more difficulty in calling TThreadList from DmtcpWorker, that's interesting too.